packages/drivers/odsp-driver/src/odspDriverUrlResolver2.ts

```diff
@@ -0,0 +1,186 @@
+import { IdentityType, TokenFetchOptions } from "./tokenFetch";
+
+export class OdspDriverUrlResolver2 implements IUrlResolver {
+    private readonly shareLinkMap = new Map<string, Promise<string>>();
+
+    public constructor(
+        private readonly getSharingLinkToken:
+            (options: TokenFetchOptions, isForFileDefaultUrl: boolean) => Promise<string | null>,
+        private readonly identityType: IdentityType = "Enterprise",
+        private readonly logger?: ITelemetryLogger,
+        private readonly appName?: string,
+    ) { }
+
+    public createCreateNewRequest(siteUrl: string, driveId: string, filePath: string, fileName: string) {
+        return createOdspCreateContainerRequest(siteUrl, driveId, filePath, fileName);
+    }
+
+    private getKey(resolvedUrl: IOdspResolvedUrl): string {
+        return `${resolvedUrl.siteUrl},${resolvedUrl.driveId},${resolvedUrl.itemId}`;
+    }
+
+    /**
+     * Takes an already generated data store url (from requestUrl) and appends a path to the
+     * existing data store information.
+     */
+    public appendDataStorePath(requestUrl: URL, pathToAppend: string): string | undefined {
+        const fluidInfo = getLocatorFromOdspUrl(requestUrl);
+
+        if (!fluidInfo) {
+            return undefined;
+        }
+
+        fluidInfo.dataStorePath = `${fluidInfo.dataStorePath}/${pathToAppend}`;
+        storeLocatorInOdspUrl(requestUrl, fluidInfo);
+
+        return requestUrl.href;
+    }
+
+    /**
+     * Resolves request URL into driver details
+     */
+    public async resolve(request: IRequest): Promise<IOdspResolvedUrl> {
+        const requestToBeResolved = { headers: request.headers, url: request.url };
+        let sharingLink: string | undefined;
+        try {
+            const url = new URL(request.url);
+            // Check if the url is the sharing link.
+            if (request.headers?.[SharingLinkHeader.isSharingLink]) {
+                sharingLink = request.url.split("?")[0];
+            }
+            const odspFluidInfo = getLocatorFromOdspUrl(url);
+            if (odspFluidInfo) {
+                requestToBeResolved.url = createOdspUrl(
+                    odspFluidInfo.siteUrl,
+                    odspFluidInfo.driveId,
+                    odspFluidInfo.fileId,
+                    odspFluidInfo.dataStorePath,
+                );
+            }
+        } catch {
+            // If the locator throws some error, then try to resolve the request as it is.
+        }
+
+        const odspResolvedUrl = await resolveRequest(requestToBeResolved);
+
+        // Generate sharingLink only if specified in the request.
+        if (requestToBeResolved.headers?.[SharingLinkHeader.generateSharingLink]) {
+            await this.getShareLinkPromise(odspResolvedUrl)
+            .then((shareLink: string) => sharingLink = shareLink)
+            .catch(() => {});
+        }
+        if (sharingLink) {
+            odspResolvedUrl.sharingLink = sharingLink;
```

Comment:
Do we have to put this on the resolved url if we have the map? We used to do this, but switched it out for the map so if it's back on the resolved url we probably don't need it. @aboktor

Reply:
We need to add it in the resolved url because resolvber is not accessible in storage service. So we need to get it off from resolved url.

Reply:
Yep! But since you're doing that anyways, I don't think you need to keep the map around in this resolver. You have access to IResolvedUrl in getAbsoluteUrl so you should already have access to the source of truth here.

Reply:
Removed caching.